add PID controller to control joint using effort

chameau5050 opened the pull request:
## Change : 

- Added PID Implementation
- Added effort-based Position Controller
- Added effort-based Velocity Controller

##  Motivation : 

The primary goal behind these changes is to model the behavior of a low-level joint controller more accurately, rather than relying on workarounds within the physics engine. Additionally, addressing some limitations related to closed-loop kinematic chains using Position and Velocity control interfaces is another objective. The PR could maybe also solve this [issue #240](https://github.com/ros-controls/gazebo_ros2_control/issues/240) 

## Change in usage : 

The usage remains largely the same as before, with one key difference. If you declare parameters kp, ki, or kd using the command_interface for Position or Velocity, the controller will apply effort to the joint based on the specified PID gains instead of directly setting the joint position or velocity.

## Usage exemple : 

Here’s an example of the new usage for controlling the velocity of a joint using an Effort PID:

```xml
<!-- define control interface -->
    <ros2_control name="my_GazeboSystem" type="system">
    
      <hardware>
        <plugin>gazebo_ros2_control/GazeboSystem</plugin>
      </hardware>
      
      <joint name="my_joint">
        
        <param name="kp">10</param>
        <param name="ki">2</param>
        <param name="kd">0.1</param>
        <param name="max_integral_error">1000</param>


        <!-- define command interface -->
        <command_interface name="velocity">
          <param name="min">0</param>
          <param name="max">2</param>
        </command_interface>

        <!-- define state interface -->
        <state_interface name="position">
          <param name="initial_value">0</param>
        </state_interface>

        <state_interface name="velocity"/>
        <state_interface name="effort"/>

      </joint>

    </ros2_control>
```

The second example retains the original behavior, where a Velocity controller sets the velocity of the joint in the physics engine :

```xml
<!-- define control interface -->
    <ros2_control name="my_GazeboSystem" type="system">
    
      <hardware>
        <plugin>gazebo_ros2_control/GazeboSystem</plugin>
      </hardware>
      
      <joint name="my_joint">
        
        <!-- define command interface -->
        <command_interface name="velocity">
          <param name="min">0</param>
          <param name="max">2</param>
        </command_interface>

        <!-- define state interface -->
        <state_interface name="position">
          <param name="initial_value">0</param>
        </state_interface>

        <state_interface name="velocity"/>
        <state_interface name="effort"/>

      </joint>

    </ros2_control>
```

ahcorde reviewed (changes requested):
can you use the PID from https://github.com/ros-controls/control_toolbox ? 
then I suggest you to add new control modes: POSITION_PID, VELOCITY_PID and EFFORT_PID, and keep the behaviour of the current controllers. This is important just in case we want to backport this feature to an older version

chameau5050 commented:
> can you use the PID from https://github.com/ros-controls/control_toolbox ? then I suggest you to add new control modes: POSITION_PID, VELOCITY_PID and EFFORT_PID, and keep the behaviour of the current controllers. This is important just in case we want to backport this feature to an older version

Sure, I can use the control_toolbox PID. I am not sure how you want me to add new control modes in terms of usage. Maybe something like that ?

```xml
<!-- define control interface -->
    <ros2_control name="my_GazeboSystem" type="system">
    
      <hardware>
        <plugin>gazebo_ros2_control/GazeboSystem</plugin>
      </hardware>
      
      <joint name="my_joint">
    
        <!-- define command interface -->
        <command_interface name="velocity_PID">
          <param name="kp">10</param>
          <param name="ki">2</param>
          <param name="kd">0.1</param>
          <param name="max_integral_error">1000</param>
        </command_interface>

        <!-- define state interface -->
        <state_interface name="position">
          <param name="initial_value">0</param>
        </state_interface>

        <state_interface name="velocity"/>
        <state_interface name="effort"/>

      </joint>

    </ros2_control>
```

christophfroehlich commented:
Yes I think this is what @ahcorde suggests. Don't forget to register the handles in `registerJoints` and the control bits in `perform_command_mode_switch` for the new tags.

Please also add a section to the documentation, and maybe also a demo to the demos package using this new feature.

chameau5050 commented:
@ahcorde @christophfroehlich, I have updated the PID implementation and added the new control mode. Please let me know if you notice any issues with this implementation. I will also update the documentation and provide examples during this week.

christophfroehlich reviewed (changes requested):
Looks good so far. Waiting for examples and documentation.

with gz_ros2_control there is a [position_proportional_gain_](https://github.com/ros-controls/gz_ros2_control/blob/9853c080b0e8ed924a65a8a8b8af8ba29d428393/gz_ros2_control/src/gz_system.cpp#L182C10-L182C37) factor for a position-controlled joint with velocity interface. This would be easy to include here, is there a use-case for this?

chameau5050 commented on gazebo_ros2_control/src/gazebo_system.cpp:
```diff
@@ -489,34 +608,47 @@ GazeboSystem::perform_command_mode_switch(
-        hardware_interface::HW_IF_VELOCITY))
+        this->dataPtr->joint_control_methods_[j] &= static_cast<ControlMethod_>(VELOCITY & VELOCITY_PID & EFFORT);
+      } 
+      else if (interface_name == (this->dataPtr->joint_names_[j] + "/" + hardware_interface::HW_IF_VELOCITY))
```
Sorry I was not familiar with the your lint system it should be fix now

christophfroehlich reviewed (changes requested):
Please run[ pre-commit locall](https://github.com/ros-controls/ros2_control/blob/master/CONTRIBUTING.md#contributing-via-pull-requests)y and also colcon test (see flake8 errors).

ahcorde commented on gazebo_ros2_control/include/gazebo_ros2_control/gazebo_system.hpp:
```diff
@@ -16,12 +16,17 @@
 
 #include "angles/angles.h"
 
+#include "control_toolbox/pid.hpp"
```
```suggestion
#include <control_toolbox/pid.hpp>
```

ahcorde commented on gazebo_ros2_control/src/gazebo_system.cpp:
```diff
@@ -83,6 +83,18 @@ class gazebo_ros2_control::GazeboSystemPrivate
   std::vector<double> joint_effort_cmd_;
 
+  /// \brief Joint velocity PID utils
+  std::vector<control_toolbox::Pid> vel_pid;
```
add include `#include <control_toolbox/pid.hpp>`

ahcorde commented on gazebo_ros2_control/src/gazebo_system.cpp:
```diff
@@ -139,6 +151,67 @@ bool GazeboSystem::initSim(
+
+  RCLCPP_INFO_STREAM(
+    this->nh_->get_logger(),
+    "setting kp=" << kp
+                  << " ki=" << ki
+                  << " kd=" << kd
+                  << " max_integral_error=" << max_integral_error);
```
```suggestion
    "Setting kp = " << kp << "\t"
                  << " ki = " << ki << "\t"
                  << " kd = " << kd << "\t"
                  << " max_integral_error = " << max_integral_error);
```

ahcorde commented on gazebo_ros2_control/src/gazebo_system.cpp:
```diff
@@ -264,13 +341,37 @@ void GazeboSystem::registerJoints(
           joint_name + suffix,
           hardware_interface::HW_IF_POSITION,
           &this->dataPtr->joint_position_cmd_[j]);
+
```
remove this change

ahcorde commented on gazebo_ros2_control_demos/examples/example_position_pid.cpp:
```diff
@@ -0,0 +1,58 @@
+// Copyright 2020 Open Source Robotics Foundation, Inc.
```
```suggestion
// Copyright 2024 Open Source Robotics Foundation, Inc.
```

ahcorde commented on gazebo_ros2_control_demos/launch/vertical_cart_example_position_pid.launch.py:
```diff
@@ -0,0 +1,85 @@
+# Copyright 2020 Open Source Robotics Foundation, Inc.
```
```suggestion
# Copyright 2024 Open Source Robotics Foundation, Inc.
```

ahcorde commented on gazebo_ros2_control_demos/launch/vertical_cart_example_velocity_pid.launch.py:
```diff
@@ -0,0 +1,85 @@
+# Copyright 2020 Open Source Robotics Foundation, Inc.
```
```suggestion
# Copyright 2024 Open Source Robotics Foundation, Inc.
```

ahcorde reviewed (changes requested):
format CI is also failing

christophfroehlich commented:
> Sorry for the confusion with the CI I was not running pre-commit with the arg --all-files so I was not seeing all error @ahcorde

Fyi: If you use the github UI to apply the suggested changes, it would be clearer for reviewer to see if this was applied or not.

chameau5050 commented:
> > Sorry for the confusion with the CI I was not running pre-commit with the arg --all-files so I was not seeing all error @ahcorde
> 
> Fyi: If you use the github UI to apply the suggested changes, it would be clearer for reviewer to see if this was applied or not.

Sorry for that, is there any other change needed in this PR ? @christophfroehlich @ahcorde

christophfroehlich reviewed (approved):
LGTM

christophfroehlich commented on gazebo_ros2_control/src/gazebo_system.cpp:
```diff
@@ -19,6 +19,8 @@
 #include <vector>
 #include <utility>
 
+#include <control_toolbox/pid.hpp>
+
```
```suggestion
#include "control_toolbox/pid.hpp"
```

christophfroehlich commented on gazebo_ros2_control/include/gazebo_ros2_control/gazebo_system.hpp:
```diff
@@ -16,12 +16,17 @@
 
 #include "angles/angles.h"
 
+#include <control_toolbox/pid.hpp>
+
 #include "gazebo_ros2_control/gazebo_system_interface.hpp"
```
```suggestion
#include "control_toolbox/pid.hpp"
#include "gazebo_ros2_control/gazebo_system_interface.hpp"
```

ahcorde commented on doc/index.rst:
```diff
@@ -131,6 +131,58 @@ We should include:
+- ``pos_max_integral_error``: Maximum summation of the error
+
+The same definitions apply to the ``vel_*`` parameters.
+
```
do you mind to add here the command to run the examples ? Then it's good to go

chameau5050 replied:
@ahcorde done

ahcorde commented:
https://github.com/Mergifyio backport iron master

ahcorde commented:
@chameau5050 any plan to include these changes in `gz_ros2_control`?

chameau5050 commented:
> @chameau5050 any plan to include these changes in `gz_ros2_control`?

@ahcorde maybe in a few months but not in the short term
